src/js/components/Breadcrumb.js

```diff
@@ -0,0 +1,295 @@
+      );
+  }
+
+  getLastItemWidth() {
+    let lastItem = ReactDOM.findDOMNode(this).lastChild;
+    let lastItemLink = lastItem.firstChild;
+
+    return lastItemLink.scrollWidth + LAST_ITEM_OFFSET;
+  }
+
+  getCurrentNamedRoutes() {
+    let {router} = this.context;
+    let namedRoutes = {};
+    let routes = router.getCurrentRoutes();
+
+    routes.forEach(function (route) {
+      // Strip regex characters from route.path
+      let path = route.path.slice().replace(/\?/g, '').replace(/\/\//g, '/');
+      if (path.charAt(0) === '/') {
+        path = path.slice(1);
+      }
+      if (path.charAt(path.length - 1) === '/') {
+        path = path.slice(0, -1);
+      }
+      // If named path already exists, don't overwrite. Sometimes root path
+      // returns a named route match and an unnamed match.
+      if (!(path in namedRoutes) || (namedRoutes[path] == null)) {
+        namedRoutes[path] = route.name;
+      }
+    });
+
+    return namedRoutes;
+  }
+
+  shouldCollapse(availableWidth, expandedWidth) {
+    // Smooth collapse action to prevent flickering between states
+    if (this.state.collapsed) {
+      return expandedWidth >= availableWidth - COLLAPSE_BUFFER;
+    }
+
+    return expandedWidth >= availableWidth + COLLAPSE_BUFFER;
+  }
+  /**
+   * Builds a crumb or multiple crumbs from the label and matching route
+   * @param  {String} label Url path part
+   * @param  {String} route Matching route name
+   * @param {Object} routeParams Router routeParams Object
+   * @param {Bool} labelIsParam Boolean indicating if the label represents a
+   * route parameter.
+   * @return {Array} Array of crumbs containing the label, route and route
+   * params. The returned label is the text shown for the crumb. The route is
+   * the route name which will be set on the Link-to attribute. The params will
+   * also be set on the Link tag. Each returned crumb will be spliced into the
+   * resulting breadcrumb.
+   */
+  buildCrumb(label, route, routeParams, labelIsParam) {
+    // Split label on whitespace and -
+    let splitLabelRegex = /[\s-]+/;
+    let joinLabelCharacter = ' ';
+    let params = {};
+
+    if (labelIsParam) {
+      // Extract param value as label e.g. (:userID -> Foo)
+      // and add param to params.
+      //
+      // Remove colon first
+      let paramName = label.slice(1);
+      // Decode value
+      label = decodeURIComponent(routeParams[paramName]);
+      params[paramName] = label;
+    } else {
+      // Split Label on space and - then capitalize each word and join.
+      // /foo-bar/ --> Foo Bar
+      label = label.split(splitLabelRegex).map(function (word) {
+        return StringUtil.capitalize(word);
+      }).join(joinLabelCharacter);
+    }
+    // Can return multiple route objects to get rendered in crumbs.
+    // Useful for splitting a route parameter like a file path.
+    return [{
+      label,
+      route,
+      params
+    }];
+  }
+
+  renderCrumbsFromRoute() {
```

Comment:
This is actually really tricky to test and haven't found an appropriate way yet. I think it would be much easier as an integration test on the pages that implement it.